Create scripts to estimate sizes of nilrt images

mike-petersen-ni opened the pull request:
These Bash scripts allow developers to predict approximate disk usage for runmode and safemode images.

More information can be found in the added README.

Output of `./estimate-usage.sh` (tab stops at 12 columns for clarity):

    x86_64      runmode     disk footprint          911M
    x86_64      safemode    disk footprint          40M
    x86_64      safemode    kernel + ramdisk        173M
    ARMv7-A     runmode     disk footprint          151M
    ARMv7-A     safemode    disk footprint          26M
    ARMv7-A     safemode    kernel + ramdisk        108M

On a PXIe-8880, after formatting the disk via MAX, extracting the x86_64 runmode image via SSH results in 899.4M of disk space used.
On a cRIO-9067, after formatting the disk via MAX, extracting the ARMv7-A runmode image via SSH results in 146M of disk space used.

I believe the discrepancy arises from the contents of `/boot`, which the estimations script do not exclude but real targets extract to a different filesystem. The size of that directory is roughtly corresponding to the discrepancy (~10M and ~5M for x86_64 and ARMv7-A).

[AB#1978832](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/1978832)

shruthi-ravi commented:
@mike-petersen-ni  can you squash the commits where logically possible? They would not be very useful in the overall nilrt repo history as they are now. Can you also add the output of the script and compare it to the actual sizes on targets in the PR description?

Commit:
Create estimation script for runmode on x86_64

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
Create estimation script for runmode on ARMv7-A

Co-authored-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>
Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
Create estimation script for safemode on x86_64

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
Create estimation script for safemode on ARMv7-A

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
Create script to run estimation scripts in a VM

As the subscripts assume superuser privileges, and perform potentially
harmful operations, a virtual machine image has been created on a
network share for this use.

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

Commit:
Add README for the estimate-usage scripts

Signed-off-by: Mike Petersen <mike.petersen@ni.com>

shruthi-ravi commented on scripts/dev/estimate-usage/estimate-usage.x86_64.runmode.sh:
```diff
@@ -0,0 +1,27 @@
+print_fs_usage() {
+    local image="$1"
+    local fs="$(mktemp)"
+    truncate -s 1G "$fs"
```
I think you could increase this to 2G since the size is already so close to 1G. Is there an x64 target with only 1G disk space?

mike-petersen-ni replied:
Makes sense to me. Will be in next push.

shruthi-ravi commented on scripts/dev/estimate-usage/estimate-usage.armv7-a.safemode.sh:
```diff
@@ -0,0 +1,31 @@
+    local input_image="$1"
+    local image="$input_image"
+    if test ! -f "$image"; then
+	# We need to look at its filesize and read it multiple times, so if it's
```
I don't follow this. Can you explain?

mike-petersen-ni replied:
These scripts can be used as part of a pipeline (in fact the top-level script uses them that way to save some disk I/O for the VM). Unlike the others, this subscript involves reading the file size (with `du --apparent-size` a couple lines later) and its content twice (with `dumpimage`). If the image given is not a plain file (e.g. a symbolic link or FIFO) the script will fail (producing an incorrect answer for the file size, and for FIFOs it will fail with an invalid format error for the second invocation of `dumpimage`, as the first invocation consumed the input already).  
It's probably possible to use `tee` to avoid this, but I figured that'd be more difficult to read.

The other scripts are unaffected by this consideration as they only read the image once, when unpacking the tarball.

shruthi-ravi commented on scripts/dev/estimate-usage/README.md:
```diff
@@ -0,0 +1,103 @@
+# `estimate-usage` scripts
+These scripts exist to allow developers to determine approximately how much disk
```
Minor: Update commit header to `Add README for the estimate-usage scripts`. Try to keep the commit header length under 50 for all commits.

mike-petersen-ni replied:
Shortened other commit messages too.

shruthi-ravi reviewed (approved):
Good work! Thanks for your help.
